### PR TITLE
Enhances tracking of installed platforms

### LINKED ISSUE #252: Track installed cores with a dedicated index file
## Bug Report

This is for a bug afflicting the IDE, see https://github.com/arduino/Arduino/issues/8936 for details. Let's double check how the same issue impacts the CLI and whether the solution being proposed might work here as well.

### PR DESCRIPTION
**Pull request by silvanocerza**
**Please check if the PR fulfills these requirements**

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?**
Enhances tracking of installed platforms.

- **What is the current behavior?**
Platforms' information is lost if the index and its related URL are removed, this means that compilation might fail too.

* **What is the new behavior?**
Platforms' information is saved at install time in `installed.json`, the file is saved in the platform installation directory.

- **Does this PR introduce a breaking change?**
No.

* **Other information**:
Fixes #252.
@per1234 should we mention this new behaviour in the documenation?

---

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

### CODE REVIEW

**Comment by per1234** [comment]
@silvanocerza the `arduino-cli board listall output is now very odd:
```
per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (scerza/core-install-fix)
$ ./arduino-cli version
arduino-cli.exe Version: 0.0.0-git Commit: 68d37ebb

per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (scerza/core-install-fix)
$ ./arduino-cli core install arduino:avr >/dev/null

per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (scerza/core-install-fix)
$ ./arduino-cli board listall
Board Name                       FQBN
Adafruit Circuit Playground      arduino:avr:
Adafruit Circuit Playground      arduino:avr:circuitplay32u4cat
Arduino ATMegaNG                 arduino:avr:
Arduino BT                       arduino:avr:
Arduino BT                       arduino:avr:bt
Arduino Diecimila                arduino:avr:
Arduino Duemilanove or Diecimila arduino:avr:diecimila
Arduino Esplora                  arduino:avr:
Arduino Esplora                  arduino:avr:esplora
Arduino Ethernet                 arduino:avr:ethernet
Arduino Ethernet                 arduino:avr:
Arduino Fio                      arduino:avr:fio
Arduino Fio                      arduino:avr:
Arduino Gemma                    arduino:avr:
Arduino Gemma                    arduino:avr:gemma
Arduino Industrial 101           arduino:avr:chiwawa
Arduino Industrial 101           arduino:avr:
Arduino Leonardo                 arduino:avr:
Arduino Leonardo                 arduino:avr:leonardo
Arduino Leonardo ETH             arduino:avr:leonardoeth
Arduino Leonardo Ethernet        arduino:avr:
Arduino LilyPadUSB               arduino:avr:
Arduino Lilypad                  arduino:avr:
Arduino Mega                     arduino:avr:
Arduino Mega ADK                 arduino:avr:megaADK
Arduino Mega or Mega 2560        arduino:avr:mega
Arduino MegaADK                  arduino:avr:
Arduino Micro                    arduino:avr:micro
Arduino Micro                    arduino:avr:
Arduino Mini                     arduino:avr:
Arduino Mini                     arduino:avr:mini
Arduino NG or older              arduino:avr:atmegang
Arduino Nano                     arduino:avr:
Arduino Nano                     arduino:avr:nano
Arduino Pro                      arduino:avr:
Arduino Pro or Pro Mini          arduino:avr:pro
Arduino Robot Control            arduino:avr:
Arduino Robot Control            arduino:avr:robotControl
Arduino Robot Motor              arduino:avr:robotMotor
Arduino Robot Motor              arduino:avr:
Arduino Uno                      arduino:avr:uno
Arduino Uno                      arduino:avr:
Arduino Uno WiFi                 arduino:avr:unowifi
Arduino Uno WiFi                 arduino:avr:
Arduino Y├║n                      arduino:avr:yun
Arduino Y├║n                      arduino:avr:
Arduino Y├║n Mini                 arduino:avr:
Arduino Y├║n Mini                 arduino:avr:yunmini
LilyPad Arduino                  arduino:avr:lilypad
LilyPad Arduino USB              arduino:avr:LilyPadUSB
Linino One                       arduino:avr:
Linino One                       arduino:avr:one
```
The expected output:
```
per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (master)
$ ./arduino-cli version
arduino-cli.exe Version: 0.0.0-git Commit: 4bb455bd

per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (master)
$ ./arduino-cli board listall
Board Name                       FQBN
Adafruit Circuit Playground      arduino:avr:circuitplay32u4cat
Arduino BT                       arduino:avr:bt
Arduino Duemilanove or Diecimila arduino:avr:diecimila
Arduino Esplora                  arduino:avr:esplora
Arduino Ethernet                 arduino:avr:ethernet
Arduino Fio                      arduino:avr:fio
Arduino Gemma                    arduino:avr:gemma
Arduino Industrial 101           arduino:avr:chiwawa
Arduino Leonardo                 arduino:avr:leonardo
Arduino Leonardo ETH             arduino:avr:leonardoeth
Arduino Mega ADK                 arduino:avr:megaADK
Arduino Mega or Mega 2560        arduino:avr:mega
Arduino Micro                    arduino:avr:micro
Arduino Mini                     arduino:avr:mini
Arduino NG or older              arduino:avr:atmegang
Arduino Nano                     arduino:avr:nano
Arduino Pro or Pro Mini          arduino:avr:pro
Arduino Robot Control            arduino:avr:robotControl
Arduino Robot Motor              arduino:avr:robotMotor
Arduino Uno                      arduino:avr:uno
Arduino Uno WiFi                 arduino:avr:unowifi
Arduino Y├║n                      arduino:avr:yun
Arduino Y├║n Mini                 arduino:avr:yunmini
LilyPad Arduino                  arduino:avr:lilypad
LilyPad Arduino USB              arduino:avr:LilyPadUSB
Linino One                       arduino:avr:one
```

**Comment by silvanocerza** [comment]
Uhu, this one is interesting. I'll take a look, thanks @per1234.

**Comment by silvanocerza** [comment]
I don't, I get the correct `ú`.

**Comment by per1234** [comment]
>do you both get that weird character for the Yùn?

I only get it on Windows Bash. I don't get it on Windows cmd or Bash on my Linux machine. However, Windows Bash can print this character:
```
per@HAL MINGW64 ~/go/src/github.com/arduino/arduino-cli (master)
$ echo Arduino Yún
Arduino Yún
```

**Comment by ubidefeo** [comment]
@per1234 it is a confusing world 😀

**Comment by silvanocerza** [comment]
Should have fixed the issue reported by @per1234, I've added a test for it too.

**Review by ubidefeo — APPROVED** [review]
I have tried removing cores, json files, entries from the config and mix and matching such occurrences without being able to reproduce the no-index issue

works for me

File: arduino/cores/packagemanager/install_uninstall_test.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+			Packager: "arduino",
+			Name:     "avrdude",
+			Version:  semver.ParseRelaxed("6.3.0-arduino14"),
+		})
+	require.Contains(t, indexPlatformRelease.ToolDependencies,
+		packageindex.IndexToolDependency{
+			Packager: "arduino",
+			Name:     "arduinoOTA",
+			Version:  semver.ParseRelaxed("1.2.1"),
+		})
```
Comment (cmaglie) [inline arduino/cores/packagemanager/install_uninstall_test.go:169]: It's ok-ish doing this way but, there must be a better way to check for matching JSONs :-)

File: commands/board/listall_test.go [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+	require.Contains(t, res.Boards, &rpc.BoardListItem{
+		Name:     "Arduino Yún",
+		FQBN:     "arduino:avr:yun",
+		IsHidden: false,
+	})
+	require.Contains(t, res.Boards, &rpc.BoardListItem{
+		Name:     "Arduino Uno",
+		FQBN:     "arduino:avr:uno",
+		IsHidden: false,
+	})
```
Comment (cmaglie) [inline commands/board/listall_test.go:75]: to reduce repetitions usually I define a local function:
```go
requireBoard := func(name, fqbn string, hidden bool) {
	require.Contains(t, res.Boards, &rpc.BoardListItem{
		Name:     name,
		FQBN:     fqbn,
		IsHidden: hidden,
	})
}
requireBoard("Arduino Yún", "arduino:avr:yun", false)
requireBoard("Arduino Uno", "arduino:avr:uno", false)
...repeat for other 25 times...
```

**Review by cmaglie — COMMENTED** [review]
Overall looks good! Nice one.

I've added some small refactoring suggestions.

I think that moving the `Board -> index.json` conversion subroutines inside the `packageindex` module may avoid us to make the index structures public and will keep them together with the subrotuine that do the opposite way (i.e. converts `index.json -> Board`).

**Comment by silvanocerza** [comment]
@cmaglie made the changes you requested, let me know if it's ok right now.

File: arduino/cores/packagemanager/install_uninstall_test.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+			require.Equal(t, expectedPlatform.Help.Online, indexPlatform.Help.Online)
+			require.Equal(t, expectedPlatform.URL, indexPlatform.URL)
+			require.Equal(t, expectedPlatform.ArchiveFileName, indexPlatform.ArchiveFileName)
+			require.Equal(t, expectedPlatform.Checksum, indexPlatform.Checksum)
+			require.Equal(t, expectedPlatform.Size, indexPlatform.Size)
+			require.ElementsMatch(t, expectedPlatform.Boards, indexPlatform.Boards)
+			require.ElementsMatch(t, expectedPlatform.ToolDependencies, indexPlatform.ToolDependencies)
+		}
+	}
+}
```
Comment (cmaglie) [inline arduino/cores/packagemanager/install_uninstall_test.go:102]: This now looks more like an integration test... I guess we can split all of this in two parts:

1. A test that verifies if the written `installed.json` contains the correct information: this one is a unit-test that should be placed in `packageindex`. In the test you can create a fake `cores.Package` and try to convert it to `json`.
2. An integration test that:
  * installs a platform and checks that the installed.json exists
  * remove the index related to the previously installed platform and checked that querying the platform will still produce the correct output (in particular `name` and `tool dependencies`)

IMHO 1. can be omitted if 2. is done propertly.

**Review by per1234 — APPROVED** [review]
It works great in all my tests Silvano. Thanks!

I was having trouble testing this previously because even before this change Arduino CLI didn't have the primary issues that inspired https://github.com/arduino/arduino-cli/issues/252. 

However, there was one condition under which I suspected Arduino CLI would still have an issue, but was unable to test because of a bug. Now that the bug has been fixed by https://github.com/arduino/arduino-cli/pull/1020, I was able to confirm that Arduino CLI does indeed misbehave under those conditions, but that this PR fixes that.

Roberto had requested that I share the procedure I used to test for this, so I give you a long-winded approval:

I created a package index for the foo vendor and saved it to `directories.data`:
```json
{
  "packages": [
    {
      "name": "foo",
      "maintainer": "Arduino",
      "websiteURL": "http://www.arduino.cc/",
      "email": "packages@arduino.cc",
      "help": {
        "online": "http://www.arduino.cc/en/Reference/HomePage"
      },
      "platforms": [
        {
          "name": "Arduino megaAVR Boards",
          "architecture": "megaavr",
          "version": "1.8.3",
          "category": "Arduino",
          "help": {
            "online": "http://www.arduino.cc/en/Reference/HomePage"
          },
          "url": "http://downloads.arduino.cc/cores/megaavr-1.8.3.tar.bz2",
          "archiveFileName": "megaavr-1.8.3.tar.bz2",
          "checksum": "SHA-256:9791d62551f1e6539769a026f903082344b8ae8b921c1796de94a2af368d866d",
          "size": "856037",
          "boards": [
            {
              "name": "Arduino Uno WiFi Rev2"
            },
            {
              "name": "Arduino Nano Every"
            }
          ],
          "toolsDependencies": [
            {
              "packager": "arduino",
              "name": "avr-gcc",
              "version": "7.3.0-atmel3.6.1-arduino5"
            },
            {
              "packager": "arduino",
              "name": "avrdude",
              "version": "6.3.0-arduino16"
            },
            {
              "packager": "arduino",
              "name": "arduinoOTA",
              "version": "1.3.0"
            }
          ]
        },
        {
          "name": "Arduino megaAVR Boards",
          "architecture": "megaavr",
          "version": "1.8.4",
          "category": "Arduino",
          "help": {
            "online": "http://www.arduino.cc/en/Reference/HomePage"
          },
          "url": "http://downloads.arduino.cc/cores/megaavr-1.8.4.tar.bz2",
          "archiveFileName": "megaavr-1.8.4.tar.bz2",
          "checksum": "SHA-256:ff25944bc6f7b766f8a71b836763517de3b12008b0b08cf24460b8af5a05abd1",
          "size": "855455",
          "boards": [
            {
              "name": "Arduino Uno WiFi Rev2"
            },
            {
              "name": "Arduino Nano Every"
            }
          ],
          "toolsDependencies": [
            {
              "packager": "arduino",
              "name": "avr-gcc",
              "version": "7.3.0-atmel3.6.1-arduino5"
            },
            {
              "packager": "arduino",
              "name": "avrdude",
              "version": "6.3.0-arduino16"
            },
            {
              "packager": "arduino",
              "name": "arduinoOTA",
              "version": "1.3.0"
            }
          ]
        },
        {
          "name": "Arduino megaAVR Boards",
          "architecture": "megaavr",
          "version": "1.8.5",
          "category": "Arduino",
          "help": {
            "online": "http://www.arduino.cc/en/Reference/HomePage"
          },
          "url": "http://downloads.arduino.cc/cores/megaavr-1.8.5.tar.bz2",
          "archiveFileName": "megaavr-1.8.5.tar.bz2",
          "checksum": "SHA-256:d35ca76fa081fb3a282584d498fadb5712a4330fa39cde560e31d72185c940d8",
          "size": "876954",
          "boards": [
            {
              "name": "Arduino Uno WiFi Rev2"
            },
            {
              "name": "Arduino Nano Every"
            }
          ],
          "toolsDependencies": [
            {
              "packager": "arduino",
              "name": "avr-gcc",
              "version": "7.3.0-atmel3.6.1-arduino5"
            },
            {
              "packager": "arduino",
              "name": "avrdude",
              "version": "6.3.0-arduino17"
            },
            {
              "packager": "arduino",
              "name": "arduinoOTA",
              "version": "1.3.0"
            }
          ]
        }
      ],
      "tools": []
    }
  ]
}
```
I then added a fake package index URL so the `package_foo_index.json` file I added to `directories.data` would be recognized:
```
export ARDUINO_BOARD_MANAGER_ADDITIONAL_URLS=https://example.com/package_foo_index.json
```
Now I install the platforms:
```
$ ./arduino-cli version
arduino-cli.exe Version: 0.0.0-git Commit: 10d07906

$ ./arduino-cli core install foo:megaavr@1.8.4
$ ./arduino-cli core install arduino:avr@1.8.2
```

Both `foo:megaavr@1.8.4` and `arduino:avr@1.8.2` have a dependency on the `arduino:avr-gcc@7.3.0-atmel3.6.1-arduino5` tool, so when I uninstall `arduino:avr`, that tool should not be removed:
```
$ ./arduino-cli core uninstall arduino:avr
Uninstalling arduino:avr@1.8.2...
arduino:avr@1.8.2 uninstalled
Uninstalling arduino:avrdude@6.3.0-arduino17, tool is no more required...
arduino:avrdude@6.3.0-arduino17 uninstalled
```
Now I remove the foo:megaavr@1.8.4 entry from its package index. This might happen if the platform author "tidies up" the package index by removing all the entries for previous versions, perhaps because they no longer wish to support them. An example is all the pre-2.3.0 versions of `esp8266:esp8266` being pulled from their package index.

Now install and uninstall `arduino:avr@1.8.2` again:
```
$ ./arduino-cli core install arduino:avr@1.8.2
./arduino-cli core uninstall arduino:avr
Uninstalling arduino:avr@1.8.2...
arduino:avr@1.8.2 uninstalled
Uninstalling arduino:avr-gcc@7.3.0-atmel3.6.1-arduino5, tool is no more required...
arduino:avr-gcc@7.3.0-atmel3.6.1-arduino5 uninstalled
Uninstalling arduino:avrdude@6.3.0-arduino17, tool is no more required...
arduino:avrdude@6.3.0-arduino17 uninstalled
Uninstalling arduino:arduinoOTA@1.3.0, tool is no more required...
arduino:arduinoOTA@1.3.0 uninstalled
```
You can see that this time the `arduino:avr-gcc@7.3.0-atmel3.6.1-arduino5` tool dependency of `foo:megaavr@1.8.4` was removed, breaking the platform.

I tried the same process again with Arduino CLI from this PR, but rebased on `master` (21cc1297) and the tool dependency is not removed when `arduino:avr` is uninstalled, leaving `foo:megaavr` intact!

**Comment by per1234** [comment]
>@per1234 should we mention this new behaviour in the documenation?

I'm thinking it's not necessary since this is almost entirely an internal implementation detail.

The only concern I had was that the `installed.json` file is placed in the platform folder, which the platform author traditionally had complete control over.

The only possible issue I can see is if there was a collision with a file of the same name placed there by the platform author. That's not very likely and it shouldn't cause any problems anyway.

**Review by cmaglie — APPROVED** [review]

**Pull request merged**